doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2310 @@
+In other words, all of the functions must treat `Self` as if it were not sized and only accessible through a pointer.[^97] A trait that can be made into a trait object is called _object safe_.
+The type `dyn Trait` always behaves as if it implemented `Trait`, which is relevant for generics, discussed below.
+
+
+#### Unsafe Traits {#unsafe-traits}
+
+It is possible to mark a trait as `unsafe` by writing `unsafe trait MyTrait { .. }`; the only difference with normal traits is that it requires `unsafe impl` to be implemented.
+Unsafe traits typically enforce some kind of additional constraint in addition to their methods; in fact, unsafe traits frequently don't have methods at all.
+For example, the standard library trait `Sync` is implemented by all types which synchronize access[^98].
+Because the invariant this trait asserts is beyond what the compiler can check, it is an unsafe trait.
+
+Trait methods may separately be marked as `unsafe`.
+This is usually done to indicate that not only does care need to be taken in implementing the trait, but calling the function also requires care (and uttering `unsafe`).
+This is separate from marking a trait as `unsafe`, and it is not necessary to mark a trait as `unsafe` for it to have `unsafe` methods.
+
+
+#### Auto Traits
+
+Auto traits are a compiler mechanism for automatically implementing certain traits; in the standard library's source code, this shows up as `auto trait Foo {}` (though this syntax is unavailable for normal libraries).
+Auto traits are implemented automatically for a struct or enum type if all of its fields also implement that trait, and are used for exposing transitive properties to the trait system.
+For example, `Send` and `Sync` are auto traits; a number of other marker traits[^99] are also auto traits.
+
+
+### Generics {#generics}
+
+_Generic programming_ is writing source code that can be compiled for many types.
+Generics are one of Rust's core features, which enable polymorphic[^100] static dispatch.
+
+Functions can be made generic by introducing _type parameters_, using a syntax similar to explicit lifetimes[^101]:
+```rust
+fn identity<T>(x: T) -> T {
+  x
+}
+```
+This function accepts a value of any type and immediately returns it.
+It can then be called like this: `identity::<i32>(42)`[^102].
+Using a generic function with all of its type parameters filled in causes it to be _instantiated_ (or _monomorphized_), resulting in code being generated for it.
+This process essentially consists of replacing each occurrence of `T` with its concrete value.
+
+Each distinct instantiation is a separate function at runtime, with a separate address, though for functions which generate identical code, like `identity::<i32>` and `identity::<u32>`, the linker may deduplicate them.
+Overzealous use of generic code can lead to binary bloat.
+
+Most of the time, the `::<>` bit is unnecessary, since Rust type deduction can infer it: `let x: u64 = identity(42);` will infer that `T = u64`.
```

Comment:
```suggestion
Most of the time, the `::<>` bit (called the `turbofish` in Rust parlance) is unnecessary, since Rust type deduction can infer it: `let x: u64 = identity(42);` will infer that `T = u64`.
```